### ISSUE TITLE
A bug in the latest version of keras: bias is not updated

### ISSUE THREAD

**Issue by bangrui** [fchollet/keras#463]
Hi guys, I have some problems with the latest version of keras. This is a simple autoencoder code:

import numpy as np

from keras.models import Sequential
from keras.layers.core import Dense, Activation, AutoEncoder
from keras.optimizers import SGD
from keras.layers import containers

train_feature = np.zeros((1,4))
for i in range(1):
        train_feature[i,] = [0.5, 0.3, 0.2, 0.1]

model = Sequential()
encoder = containers.Sequential([Dense(4,2,activation='tanh')])
decoder = containers.Sequential([Dense(2,4,activation='tanh')])
autoencoder = AutoEncoder(encoder=encoder,decoder=decoder)
model.add(autoencoder)

sgd = SGD(lr = 1)
model.compile(loss='mse', optimizer=sgd)
weights_org = model.get_weights()
model.fit(train_feature, train_feature, nb_epoch=1, batch_size=1, shuffle=False)
weights_now = model.get_weights()
print "Weights and bias before trainning is"
print weights_org
print "Weights and bias after trainning is"
print weights_now

If I use the past version, the result is fine, which looks like:
Epoch 0
1/1 [==============================] - 0s - loss: 0.1271
Weights and bias before trainning is
[array([[-0.92038186,  0.05200247],
       [ 0.46043892, -0.82942702],
       [-0.57505271, -0.25431725],
       [ 0.03477517, -0.10212894]]), array([ 0.,  0.]), array([[-0.16022001, -0.06501606, -0.68613223,  0.96541306],
       [ 0.15240795, -0.41876942,  0.50621491,  0.11618159]]), array([ 0.,  0.,  0.,  0.])]
Weights and bias after trainning is
[array([[-0.86223659,  0.07199206],
       [ 0.49532608, -0.81743327],
       [-0.5517946 , -0.24632141],
       [ 0.04640422, -0.09813102]]), array([ 0.11629053,  0.03997918]), array([[-0.25749549, -0.09675063, -0.69828947,  0.87946589],
       [ 0.08653298, -0.44026008,  0.49798202,  0.05797814]]), array([ 0.23823399,  0.07772003,  0.02977387,  0.21049024])]
However, if I use the latest version of keras, then the result will be:
Epoch 0
1/1 [==============================] - 0s - loss: 0.0328
Weights and bias before trainning is
[array([[ 0.99807841, -0.0321337 ],
       [-0.34120891,  0.31954411],
       [ 0.61708271,  0.51410343],
       [-0.68198998,  0.58714997]]), array([ 0.,  0.]), array([[ 0.69201076,  0.3782278 ,  0.95140293,  0.22793627],
       [ 0.16048078,  0.68251729,  0.26576419, -0.86654349]]), array([ 0.,  0.,  0.,  0.])]
Weights and bias after trainning is
[array([[ 0.99807841, -0.0321337 ],
       [-0.34120891,  0.31954411],
       [ 0.61708271,  0.51410343],
       [-0.68198998,  0.58714997]]), array([ 0.,  0.]), array([[ 0.69201076,  0.3782278 ,  0.95140293,  0.22793627],
       [ 0.16048078,  0.68251729,  0.26576419, -0.86654349]]), array([ 0.,  0.,  0.,  0.])]

Basically, the bias of the neural network is not updating. Could someone help to fix it?

**Comment by wuaalb** [comment]
Same thing happens with check_autoencoder.py for instance

```
...
Training Classical AutoEncoder
...
Epoch 0
5000/5000 [==============================] - 0s - loss: 0.2017
Epoch 1
5000/5000 [==============================] - 0s - loss: 0.2017
Epoch 2
5000/5000 [==============================] - 0s - loss: 0.2017
...
```

I think the problem has to do with the following:

Commit "Incorporate regularization into loss function" f04fdec9e61eabe9750af00f56c18bdde8972045 changed

1. `Dense`'s regularizers list will only contain non-`None` regularizers pass in the initializer, ie. by default the regularizers list is empty.
2. `Layer`'s `get_params()` just returns the regularizers list, instead of doing some logic which ensures there's a (identity) regularizer for each parameter.

Then commit "Remove weight tying in autoencoder" 94c930e99e8908d2188213672bed050b54ebdb5a changed AutoEncoder's initializer like
```diff
-            self.params += layer.params
-            if hasattr(layer, 'regularizers'):
-                self.regularizers += layer.regularizers
-            if hasattr(layer, 'constraints'):
-                self.constraints += layer.constraints
+            params, regularizers, constraints = layer.get_params()
+            self.constraints += constraints
+            for p, r in zip(params, regularizers):
+                if p not in self.params:
+                    self.params.append(p)
+                    self.regularizers.append(r)
```
which does a `zip(params, regularizers)` between two lists which now can have different lengths.
In the case of the default `Dense` encoder/decoder (i.e. with no regularizers), this means the parameter list will be empty and nothing will be updated during training.

See also #455.

**Comment by fchollet** [comment]
Seemed like an issue indeed. I fixed it.
